Kickstarter-iOS/Features/PledgeView/Controllers/PostCampaignCheckoutViewController.swift

```diff
@@ -402,38 +403,83 @@ extension PostCampaignCheckoutViewController: PledgeViewControllerMessageDisplay
-    didAuthorizePayment payment: PKPayment,
-    handler completion: @escaping (PKPaymentAuthorizationResult)
-      -> Void
+extension PostCampaignCheckoutViewController: STPApplePayContextDelegate {
+  func applePayContext(
+    _: StripeApplePay.STPApplePayContext,
+    didCreatePaymentMethod _: StripePayments.STPPaymentMethod,
+    paymentInformation payment: PKPayment,
+    completion: @escaping StripeApplePay.STPIntentClientSecretCompletionBlock
   ) {
-    let paymentDisplayName = payment.token.paymentMethod.displayName
-    let paymentNetworkName = payment.token.paymentMethod.network?.rawValue
+    guard let paymentIntentClientSecret = self.applePayPaymentIntent else {
+      completion(
+        nil,
+        PostCampaignCheckoutApplePayError.missingPaymentIntent("Missing PaymentIntent")
+      )
+      return
+    }
+
+    guard let paymentDisplayName = payment.token.paymentMethod.displayName,
+      let paymentNetworkName = payment.token.paymentMethod.network?.rawValue else {
+      completion(
+        nil,
+        PostCampaignCheckoutApplePayError
+          .missingPaymentMethodInfo("Unable to retrieve payment method information from ApplePay")
+      )
+      return
+    }
+
     let transactionId = payment.token.transactionIdentifier
 
-    self.viewModel.inputs.paymentAuthorizationDidAuthorizePayment(paymentData: (
-      paymentDisplayName,
-      paymentNetworkName,
-      transactionId
-    ))
+    // Tokens considered 'legacy' from Stripe's side, but we still store the token ourselves.
+    STPAPIClient.shared.createToken(with: payment) { [weak self] token, _ in
+      guard let self = self else {
+        return
+      }
+
+      guard let tokenId = token?.tokenId else {
+        completion(
+          nil,
+          PostCampaignCheckoutApplePayError.missingToken("Unable to retrieve token from Stripe")
+        )
+        return
+      }
 
-    STPAPIClient.shared.createToken(with: payment) { [weak self] token, error in
-      guard let self = self else { return }
+      let params = ApplePayParams(
+        paymentInstrumentName: paymentDisplayName,
+        paymentNetwork: paymentNetworkName,
+        transactionIdentifier: transactionId,
+        token: tokenId
+      )
 
-      let status = self.viewModel.inputs.stripeTokenCreated(token: token?.tokenId, error: error)
-      let result = PKPaymentAuthorizationResult(status: status, errors: [])
+      self.viewModel.inputs.applePayContextDidCreatePayment(params: params)
+      completion(paymentIntentClientSecret, nil)
+    }
+  }
 
-      completion(result)
+  func applePayContext(
+    _: StripeApplePay.STPApplePayContext,
+    didCompleteWith status: StripePayments.STPPaymentStatus,
+    error _: Error?
+  ) {
+    switch status {
+    case .success:
+      self.viewModel.inputs.applePayContextDidComplete()
+    case .error:
+      self.messageBannerViewController?
+        .showBanner(with: .error, message: Strings.Something_went_wrong_please_try_again())
```

Comment:
One thing that isn't completely clear to me - if Stripe errors at this point, has the payment been confirmed? I _assume_ not.

Reply:
Good question. have you tried testing with a failure test card and then looking in the stripe dashboard (with Bessie's help) to see if it went through?

Reply:
Either way, we'll need to run our validations/checkout ID calls and reconfirm again, but it would be good to know.